File: packages/agent-sdk/src/answer.ts
/**
 * AnswerBuilder: makes the compliant answer shape (text + citations +
 * confidence, abstention as a success mode) the path of least resistance.
 */

import type { Answer, Citation } from '@acp/protocol';

/**
 * Answers below this confidence must abstain rather than guess
 * (knowledge-and-rag.md); agents may only raise the floor.
 */
export const DEFAULT_CONFIDENCE_FLOOR = 0.35;

/**
 * Round-half-away-from-zero to 4 decimals. Deliberately NOT Python's
 * banker's rounding — parity fixtures avoid ties so both SDKs agree
 * (see fixtures/parity/HANDLERS.md).
 */
function round4(x: number): number {
  return Math.round(x * 1e4) / 1e4;
}

export class AnswerBuilder {
  readonly confidenceFloor: number;
  private readonly paragraphs: string[] = [];
  private readonly citations: Citation[] = [];

  constructor(confidenceFloor: number = DEFAULT_CONFIDENCE_FLOOR) {
    this.confidenceFloor = confidenceFloor;
  }

  /**
   * Registers a citation and returns its 1-based marker number.
   *
   * The same lineage_id cites once — repeated grounding on one chunk should
   * not inflate the citation list.
   */
  cite(citation: Citation): number {
    for (const [i, existing] of this.citations.entries()) {
      if (existing.lineage_id === citation.lineage_id) return i + 1;
    }
    this.citations.push(citation);
    return this.citations.length;
  }

  paragraph(text: string): void {
    this.paragraphs.push(text.trim());
  }

  /** Below the floor this abstains; on success `abstained` is omitted entirely. */
  build(confidence: number): Answer {
    if (confidence < this.confidenceFloor) {
      return this.abstain(
        "I don't have sufficient grounding in the corpus to answer this reliably.",
        confidence,
      );
    }
    return {
      text: this.paragraphs.join('\n\n'),
      citations: this.citations,
      confidence: round4(confidence),
    };
  }

  /** Abstention is a success mode, never a confident guess. */
  abstain(reason: string, confidence = 0): Answer {
    return {
      text: reason,
      citations: [],
      confidence: round4(Math.min(confidence, this.confidenceFloor)),
      abstained: true,
    };
  }
}

File: python/acp-agent-sdk/src/acp_agent_sdk/answer.py
"""AnswerBuilder: makes the compliant answer shape (text + citations +
confidence, abstention as a success mode) the path of least resistance."""

from dataclasses import dataclass, field
from typing import Any

# Answers below this confidence must abstain rather than guess
# (knowledge-and-rag.md); agents may only raise the floor.
DEFAULT_CONFIDENCE_FLOOR = 0.35


@dataclass
class AnswerBuilder:
    confidence_floor: float = DEFAULT_CONFIDENCE_FLOOR
    _paragraphs: list[str] = field(default_factory=list)
    _citations: list[dict[str, Any]] = field(default_factory=list)

    def cite(self, citation: dict[str, Any]) -> int:
        """Registers a citation and returns its 1-based marker number.

        The same lineage_id cites once — repeated grounding on one chunk
        should not inflate the citation list.
        """
        for i, existing in enumerate(self._citations):
            if existing.get("lineage_id") == citation.get("lineage_id"):
                return i + 1
        self._citations.append(citation)
        return len(self._citations)

    def paragraph(self, text: str) -> None:
        self._paragraphs.append(text.strip())

    def build(self, confidence: float) -> dict[str, Any]:
        if confidence < self.confidence_floor:
            return self.abstain(
                "I don't have sufficient grounding in the corpus to answer this reliably.",
                confidence,
            )
        return {
            "text": "\n\n".join(self._paragraphs),
            "citations": self._citations,
            "confidence": round(confidence, 4),
        }

    def abstain(self, reason: str, confidence: float = 0.0) -> dict[str, Any]:
        """Abstention is a success mode, never a confident guess."""
        return {
            "text": reason,
            "citations": [],
            "confidence": round(min(confidence, self.confidence_floor), 4),
            "abstained": True,
        }
